src/Media/Ingester/SideloadDir.php

```diff
@@ -0,0 +1,409 @@
+        }
+
+        $select = new Element\Select('o:media[__index__][ingest_folder]');
+        $select
+            ->setOptions([
+                'label' => 'Folder', // @translate
+                'info' => 'Folders and files without sufficient permissions are skipped.', // @translate
+                'value_options' => $this->listDirs,
+                'empty_option' => $emptyOptionDir,
+            ])
+            ->setAttributes([
+                'id' => 'media-sideload-ingest-folder-__index__',
+                'required' => true,
+            ]);
+
+        $recursive = new Element\Checkbox('o:media[__index__][ingest_folder_recursively]');
+        $recursive
+            ->setOptions([
+                'label' => 'Ingest folder recursively', // @translate
+            ])
+            ->setAttributes([
+                'id' => 'media-sideload-ingest-folder-recursive-__index__',
+                'required' => false,
+            ]);
+
+        return $view->formRow($select)
+            . $view->formRow($recursive);
+    }
+
+    /**
+     * Get all folders available to sideload.
+     */
+    protected function listDirs(): void
+    {
+        $this->listDirs = [];
+        $this->hasMoreDirs = false;
+
+        $dir = new \SplFileInfo($this->directory);
+        if (!$dir->isDir()) {
+            return;
+        }
+
+        $countDirs = 0;
+
+        $lengthDir = strlen($this->directory) + 1;
+        $dir = new \RecursiveDirectoryIterator($this->directory);
+        // Prevent UnexpectedValueException "Permission denied" by excluding
+        // directories that are not executable or readable.
+        $dir = new \RecursiveCallbackFilterIterator($dir, function ($current, $key, $iterator) {
+            if ($iterator->isDir() && (!$iterator->isExecutable() || !$iterator->isReadable())) {
+                return false;
+            }
+            return true;
+        });
+        $iterator = new \RecursiveIteratorIterator($dir);
+        /** @var \SplFileInfo $file */
+        foreach ($iterator as $filepath => $file) {
+            if ($file->isDir()) {
+                if (!$this->hasMoreDirs && $this->verifyFileOrDir($file, true)) {
+                    // There are two filepaths for one dirpath: "." and "..".
+                    $filepath = $file->getRealPath();
+                    // Don't list empty directories.
+                    // TODO Add the count of files for display.
+                    if (!$this->dirHasNoFileAndIsRemovable($filepath)) {
+                        // For security, don't display the full path to the user.
+                        $relativePath = substr($filepath, $lengthDir);
+                        if (!isset($this->listDirs[$relativePath])) {
+                            // Use keys for quicker process on big directories.
+                            $this->listDirs[$relativePath] = null;
+                            if ($this->maxFiles && ++$countDirs >= $this->maxFiles) {
```

Comment:
I'm not sure what's happening but this seems to ignore the $maxFiles configuration. I set my max files to 10 but all files were ingested by this ingester.

Reply:
Or another option if needed.

Reply:
Added in last commit.